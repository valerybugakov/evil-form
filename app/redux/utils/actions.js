/* flow */
import { createAction } from 'redux-act'
import { identity, noop } from 'lodash/fp'
import { fakeStore } from 'redux/utils/dispatch'

export const INIT = 'init'
export const SUCCESS = 'success'
export const FAILURE = 'failure'

const requestMetaReducer = (actionName, {
  serializer,
  deserializer,
  applySerializerToPayload,
  ...rest
}) => ((payload, resolve = noop, reject = noop) => ({
  resolve,
  reject,
  serializer,
  deserializer,
  type: actionName,
  requestStatus: INIT,
  applySerializerToPayload,
  requestOptions: typeof rest.requestOptions === 'function'
    ? rest.requestOptions(payload)
    : rest,
}))

const createRequestStatusAction = (actionName, status) => (
  createAction(`${actionName} ${status}`, identity, (_, payload) => ({
    type: actionName,
    requestStatus: status,
    ...payload,
  }))
)

export const createRequestActions = ({
  desc: actionName,
  ...metaData
}) => {
  const res = {
    id: actionName,
    [INIT]: createAction(
      `${actionName} ${INIT}`,
      identity,
      requestMetaReducer(actionName, metaData),
    ).assignTo(fakeStore),
  }

  return [SUCCESS, FAILURE].reduce((reqs, type) => {
    reqs[type] = createRequestStatusAction(actionName, type)
    return reqs
  }, res)
}

export const createActionBinded = (value, process, meta) => (
  createAction(value, process, meta).assignTo(fakeStore)
)

export const promisifyAction = (action, payload) => (
  new Promise((resolve, reject) => action(payload, resolve, reject))
)

export function actionsGenerator(actionConfig, path = []) {
  const minimumActionConfig = {
    desc: path.join('.'),
  }

  // Create ordinary action if null provided as config
  if (actionConfig === null) {
    return createActionBinded(minimumActionConfig.desc)
  }

  // Create requestActions if requestOptions or url provided in config
  if (actionConfig.requestOptions || actionConfig.url) {
    return createRequestActions({
      ...minimumActionConfig,
      ...actionConfig,
    })
  }

  // Recursively call actionsGenerator on all config keys
  if (typeof actionConfig === 'object') {
    return Object.entries(actionConfig).reduce((acc, [key, value]) => {
      acc[key] = actionsGenerator(value, [...path, key])
      return acc
    }, {})
  }

  return actionConfig
}
