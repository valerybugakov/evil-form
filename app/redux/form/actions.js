import { actionsGenerator } from 'redux/utils'

export default actionsGenerator({
  test: null,
})

