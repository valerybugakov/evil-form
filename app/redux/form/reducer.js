import { createReducer } from 'redux-act'
import { reducer as formReducer } from 'redux-form'
import actions from './actions'

const initialState = {
  values: {
    title: 'San Francisco Driver Form',
    description: 'Welcome aboard!',
    fields: [
      {
        type: 'text',
        title: 'SSN',
        required: true,
      },
      {
        type: 'radio',
        title: 'Have you driven a car before?',
      },
      {
        type: 'checkbox',
        title: 'Where do you want to work?',
      },
      {
        type: 'select',
      },
      {
        type: 'file',
      },
      {
        type: 'textarea',
      },
    ],
  },
}

const formBuilder = createReducer({
  [actions.test](state) {
    return state
  },
}, initialState)

export default formReducer.plugin({
  formBuilder,
})
