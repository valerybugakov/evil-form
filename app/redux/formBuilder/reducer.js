import { createReducer } from 'redux-act'
import { saveFormSuccess } from './actions'

const initialState = {
  savedForms: [],
}

export default createReducer({
  [saveFormSuccess]: (state, savedForm) => ({
    ...state,
    savedForms: [...state.savedForms, savedForm],
  }),
}, initialState)
