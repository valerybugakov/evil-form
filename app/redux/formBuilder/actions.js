import { createActionBinded } from 'redux/utils'

export const saveForm = createActionBinded('Save form init')
export const saveFormSuccess = createActionBinded('Save form success')
export const saveFormFailure = createActionBinded('Save form failure')
