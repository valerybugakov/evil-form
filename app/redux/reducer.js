import { combineReducers } from 'redux'
import { reducer as form } from 'redux-form'
import formBuilder from './formBuilder/reducer'

export default combineReducers({
  form,
  formBuilder,
})
