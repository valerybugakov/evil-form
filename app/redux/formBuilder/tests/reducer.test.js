import { saveFormSuccess } from '../actions'
import reducer from '../reducer'

describe('formBuidler reducer', () => {
  const existingForm = { id: 1, title: 'Test form' }
  const initialState = { savedForms: { 1: existingForm } }

  it('saves new form not matched by ID', () => {
    const newForm = { id: 2, title: 'New title' }
    const next = reducer(initialState, saveFormSuccess(newForm))
    expect(next).toMatchSnapshot()
  })

  it('updates existing form matched by ID', () => {
    const updatedForm = { id: 1, title: 'New title' }
    const next = reducer(initialState, saveFormSuccess(updatedForm))
    expect(next).toMatchSnapshot()
  })
})
