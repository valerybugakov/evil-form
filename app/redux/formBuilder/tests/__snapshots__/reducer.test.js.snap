// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formBuidler reducer saves new form not matched by ID 1`] = `
Object {
  "savedForms": Object {
    "1": Object {
      "id": 1,
      "title": "Test form",
    },
    "2": Object {
      "id": 2,
      "title": "New title",
    },
  },
}
`;

exports[`formBuidler reducer updates existing form matched by ID 1`] = `
Object {
  "savedForms": Object {
    "1": Object {
      "id": 1,
      "title": "New title",
    },
  },
}
`;
